git/options.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2021 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"fmt"
+	"net/url"
+)
+
+const (
+	DefaultRemote            = "origin"
+	DefaultBranch            = "master"
+	DefaultPublicKeyAuthUser = "git"
+)
+
+// CheckoutOptions are the options used for a Git checkout.
+type CheckoutOptions struct {
+	// Branch to checkout, can be combined with Branch with some
+	// Implementations.
+	Branch string
+
+	// Tag to checkout, takes precedence over Branch.
+	Tag string
+
+	// SemVer tag expression to checkout, takes precedence over Tag.
+	SemVer string `json:"semver,omitempty"`
+
+	// Commit SHA1 to checkout, takes precedence over Tag and SemVer,
+	// can be combined with Branch with some Implementations.
+	Commit string
+
+	// RecurseSubmodules defines if submodules should be checked out,
+	// not supported by all Implementations.
+	RecurseSubmodules bool
```

Comment:
`RecurseSubmodules` and `ShallowClone` could be extracted into a implementation dependent state machine field - not necessarily as part of this PR.